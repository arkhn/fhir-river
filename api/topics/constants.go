package topics

const (
	BatchPrefix     = "batch."
	ExtractPrefix   = "extract."
	TransformPrefix = "transform."
	LoadPrefix      = "load."
	NumParts        = 3
	Batch           = "^batch.*"
	Extract         = "^extract.*"
	Transform       = "^transform.*"
	Load            = "^load.*"
)
