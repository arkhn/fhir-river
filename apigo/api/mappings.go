package api

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
)

const resourceFromIDQuery = `
fragment entireFilter on Filter {
    id
    sqlColumn {
        id
        table
        column
    }
    relation
    value
}

fragment entireColumn on Column {
    table
    column
    joins {
        tables {
            table
            column
        }
    }
}

fragment entireInput on Input {
    sqlValue {
        ...entireColumn
    }
    script
    conceptMapId
    staticValue
}

fragment entireCondition on Condition {
    action
    sqlValue {
        table
        column
    }
    relation
    value
}

fragment entireInputGroup on InputGroup {
    id
    mergingScript
    inputs {
        ...entireInput
    }
    conditions {
        ...entireCondition
    }
}

fragment a on Attribute {
    path
    definitionId
    inputGroups {
        ...entireInputGroup
    }
}

fragment cred on Credential {
    model
    host
    port
    database
    owner
    login
    password: decryptedPassword
}

query resource($resourceId: ID!) {
    resource(resourceId: $resourceId) {
        id
        primaryKeyTable
        primaryKeyColumn
        definitionId
        definition {
            type
            kind
            derivation
            url
        }
        filters {
            ...entireFilter
        }
        attributes {
            ...a
        }
        source {
            id
            credential {
                ...cred
            }
        }
    }
}`

type graphqlVariables struct {
	ResourceID string `json:"resourceId"`
}

type mappingCredentials struct {
	Model    string `json:"model"`
	Host     string `json:"host"`
	Port     string `json:"port"`
	Database string `json:"database"`
	Owner    string `json:"owner"`
	Login    string `json:"login"`
	Password string `json:"password"`
}

type mappingColumn struct {
	Table  string `json:"table"`
	Column string `json:"column"`
}

type mappingFilter struct {
	ID        string        `json:"id"`
	SqlColumn mappingColumn `json:"sqlColumn"`
	Relation  string        `json:"relation"`
	Value     string        `json:"value"`
}

type mappingAttribute struct {
	Path         string              `json:"path"`
	DefinitionId string              `json:"definitionId"`
	InputGroups  []mappingInputGroup `json:"inputGroups"`
}

type mappingInputGroup struct {
	ID            string             `json:"id"`
	MergingScript string             `json:"mergingScript"`
	Inputs        []mappingInput     `json:"inputs"`
	Conditions    []mappingCondition `json:"conditions"`
}

type mappingInput struct {
	SqlValue struct {
		Table  string `json:"table"`
		Column string `json:"column"`
		Joins  []struct {
			Tables []mappingColumn `json:"tables"`
		} `json:"joins"`
	} `json:"sqlValue"`
	Script          string            `json:"script"`
	ConceptMapID    string            `json:"conceptMapId"`
	ConceptMapTitle string            `json:"conceptMapTitle"`
	ConceptMap      map[string]string `json:"conceptMap"`
	StaticValue     string            `json:"staticValue"`
}

type mappingCondition struct {
	Action   string        `json:"action"`
	SqlValue mappingColumn `json:"sqlValue"`
	Relation string        `json:"relation"`
	Value    string        `json:"value"`
}

type mappingResource struct {
	ID               string `json:"id"`
	PrimaryKeyTable  string `json:"primaryKeyTable"`
	PrimaryKeyColumn string `json:"primaryKeyColumn"`
	DefinitionId     string `json:"definitionId"`
	Definition       struct {
		Type       string `json:"type"`
		Kind       string `json:"kind"`
		Derivation string `json:"derivation"`
		Url        string `json:"url"`
	} `json:"definition"`
	Filters    []mappingFilter    `json:"filters"`
	Attributes []mappingAttribute `json:"attributes"`
	Source     struct {
		ID         string             `json:"id"`
		Credential mappingCredentials `json:"credential"`
	} `json:"source"`
}

type graphqlResponse struct {
	Data struct {
		Resource mappingResource `json:"resource"`
	} `json:"data"`
}

func fetchMapping(resourceID string, authorizationHeader string) (*mappingResource, error) {
	// TODO make a type for the response?
	variables := graphqlVariables{
		ResourceID: resourceID,
	}
	graphqlQueryBody := struct {
		Query     string           `json:"query"`
		Variables graphqlVariables `json:"variables"`
	}{
		Query:     resourceFromIDQuery,
		Variables: variables,
	}
	jBody, _ := json.Marshal(graphqlQueryBody)
	graphqlQuery, err := http.NewRequest("POST", pyrogURL, bytes.NewBuffer(jBody))
	if err != nil {
		return nil, err
	}
	graphqlQuery.Header.Set("Authorization", authorizationHeader)
	graphqlQuery.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(graphqlQuery)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	gqlResp := graphqlResponse{}
	err = json.NewDecoder(resp.Body).Decode(&gqlResp)
	if err != nil {
		return nil, err
	}

	resourceMapping := gqlResp.Data.Resource

	// Dereference concept maps
	err = dereferenceConceptMap(&resourceMapping, authorizationHeader)
	if err != nil {
		return nil, err
	}

	return &resourceMapping, nil
}

func dereferenceConceptMap(mapping *mappingResource, authorizationHeader string) error {
	for _, attribute := range mapping.Attributes {
		for _, inputGroup := range attribute.InputGroups {
			for _, input := range inputGroup.Inputs {
				if input.ConceptMapID != "" {
					conceptMap, err := fetchConceptMap(input.ConceptMapID, authorizationHeader)
					if err != nil {
						return err
					}
					input.ConceptMap = conceptMap
				}
			}
		}
	}
	return nil
}

type conceptMapResponse struct {
	Group []struct {
		Element []struct {
			Code   string `json:"code"`
			Target []struct {
				Equivalence string `json:"equivalence"`
				Code        string `json:"code"`
			} `json:"target"`
		} `json:"element"`
	} `json:"group"`
}

func fetchConceptMap(conceptMapID string, authorizationHeader string) (map[string]string, error) {
	url := fmt.Sprintf("%s/ConceptMap/%s", fhirURL, conceptMapID)
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Authorization", authorizationHeader)
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}

	switch resp.StatusCode {
	case 200:
		// If everything went well, we go on
	case 401:
		return nil, &invalidTokenError{message: "Token is invalid", statusCode: 401}
	case 403:
		return nil, &invalidTokenError{message: "You don't have rights to perform this action", statusCode: 403}
	default:
		// Return other errors
		return nil, errors.New("Error while fetching concept map")
	}

	conceptMapResp := conceptMapResponse{}
	err = json.NewDecoder(resp.Body).Decode(&conceptMapResp)
	if err != nil {
		return nil, err
	}

	// Turn the concept map into a map
	conceptMap := make(map[string]string)
	for _, group := range conceptMapResp.Group {
		for _, element := range group.Element {
			// NOTE we only handle a single target for each source
			sourceCode := element.Code
			targetCode := element.Target[0].Code
			conceptMap[sourceCode] = targetCode
		}
	}

	return conceptMap, nil
}

func storeMapping(mapping []byte, resourceID string, batchID string) error {
	err := rdb.Set(fmt.Sprintf("%s:%s", batchID, resourceID), mapping, 0)
	if err != nil {
		return err.Err()
	}
	return nil
}
