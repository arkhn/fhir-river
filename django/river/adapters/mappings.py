from typing import Any

from common.mapping.fetch_mapping import fetch_resource_mapping
from utils.caching import cache


class MappingUnavailable(Exception):
    pass


def build_repository_key(batch_id: str, resource_id: str):
    return f"{batch_id}:{resource_id}"


class MappingsRepository:
    """This class is an abstraction that is used when we need to fetch a mapping.
    It has a single method `get` with the mapping id as argument.
    """

    def set(self, batch_id: str, resource_id: str, mapping: Any):
        raise NotImplementedError

    def get(self, batch_id: str, resource_id: str):
        raise NotImplementedError


class FakeMappingsRepository(MappingsRepository):
    def __init__(self, mappings: dict = None):
        self._mappings = mappings or {}
        self._seen = []

    def set(self, batch_id: str, resource_id: str, mapping: Any):
        key = build_repository_key(batch_id, resource_id)
        self._mappings[key] = mapping

    def get(self, batch_id: str, resource_id: str):
        key = build_repository_key(batch_id, resource_id)
        self._seen.append(ikeyd)
        return self._mappings[key]


class RedisMappingsRepository(MappingsRepository):
    def __init__(self, mapping_redis: redis.Redis):
        self.mapping_redis = mapping_redis

    def set(self, batch_id: str, resource_id: str, mapping: Any):
        key = build_repository_key(batch_id, resource_id)
        self.mappings_redis.set(key, json.dumps(resource_mapping))

    def get(self, batch_id: str, resource_id: str):
        key = build_repository_key(batch_id, resource_id)
        serialized_mapping = self.mapping_redis.get(key)
        if serialized_mapping is None:
            logger.exception(
                {
                    "message": f"Mapping not found for batch {batch_id} and resource {resource_id}",
                    "resource_id": resource_id,
                },
            )
            raise MappingUnavailable(resource_id)
        mapping = json.loads(serialized_mapping)

        return mapping
