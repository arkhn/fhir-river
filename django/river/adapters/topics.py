import logging

from confluent_kafka import admin, error

from django.conf import settings

logger = logging.getLogger(__name__)


class TopicsManager:
    def create(self, topic: str):
        raise NotImplementedError

    def delete(self, topic: str):
        raise NotImplementedError


class FakeTopicsManager(TopicsManager):
    def __init__(self, topics=list()):
        self._topics = set(topics)

    def create(self, topic: str):
        self._topics.add(topic)

    def delete(self, topic: str):
        self._topics.remove(topic)


class KafkaTopicsManager(TopicsManager):
    def __init__(self):
        self._admin_client = admin.AdminClient({"bootstrap.servers": settings.KAFKA_BOOTSTRAP_SERVERS})

    def create(self, topic: str):
        future_topics = self._admin_client.create_topics(
            [admin.NewTopic(topic, settings.KAFKA_NUM_PARTITIONS, settings.KAFKA_REPLICATION_FACTOR)]
        )
        logger.debug(f"List of kafka topics after addition: {self._admin_client.list_topics().topics}")
        try:
            future_topics[topic].result(10)
        except error.KafkaException as exc:
            raise Exception(exc)

    def delete(self, topic: str):
        future_deleted_topics = self._admin_client.delete_topics([topic])
        logger.debug(f"List of kafka topics after deletion: {self._admin_client.list_topics().topics}")
        try:
            future_deleted_topics[topic].result(10)
        except error.KafkaException as exc:
            logger.error(f"KafkaException while deleting topics: {str(exc)}")
