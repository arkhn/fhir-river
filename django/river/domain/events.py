from dataclasses import dataclass
from typing import Any, Dict, List, Optional


class Event:
    pass


@dataclass(frozen=True)
class BatchEvent(Event):
    batch_id: str
    resource_id: str
    primary_key_values: Optional[List[str]] = None


@dataclass(frozen=True)
class ExtractedRecord(Event):
    batch_id: str
    resource_type: str
    resource_id: str
    record: Dict[str, Any]
