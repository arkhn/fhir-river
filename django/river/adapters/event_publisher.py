import dataclasses
import json
import logging

import confluent_kafka

from django.conf import settings

from river.domain.events import Event

logger = logging.getLogger(__name__)


class EventPublisher:
    def publish(self, topic: str, event: Event):
        raise NotImplementedError


class FakeEventPublisher(EventPublisher):
    def __init__(self):
        self._events: dict[str, set] = {}

    def publish(self, topic: str, event: Event):
        if topic in self._events:
            self._events[topic].add(event)
        else:
            self._events[topic] = {event}


class KafkaEventPublisher(EventPublisher):
    def __init__(
        self,
    ):
        self._kafka_producer = confluent_kafka.Producer(
            {
                "bootstrap.servers": settings.KAFKA_BOOTSTRAP_SERVERS,
                "linger.ms": 0.5,
                "session.timeout.ms": 6000,
            }
        )

    def publish(self, topic: str, event: Event):
        self._kafka_producer.produce(
            topic=topic,
            value=json.dumps(dataclasses.asdict(event)),
            callback=lambda err, msg: logger.debug("Message delivered")
            if err is None
            else logger.error(f"Message {event} delivery failed with error {err} for topic {msg.topic()}"),
        )
        self._kafka_producer.poll(1)
