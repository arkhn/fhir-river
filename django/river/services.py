import json
from typing import Any, List, Optional, Tuple

from django.utils import timezone

from fhir.resources import construct_fhir_element

from pydantic import ValidationError
from river import models
from river.adapters.event_publisher import EventPublisher
from river.adapters.pyrog_client import PyrogClient
from river.adapters.topics import TopicsManager
from river.common.analyzer import Analyzer
from river.common.database_connection.db_connection import DBConnection
from river.domain.events import BatchEvent
from river.extractor.extractor import Extractor
from river.transformer.transformer import Transformer
from utils.json import CustomJSONEncoder


def build_mappings(
    mappings,
    # FIXME remove this when the DB is shared
    pyrog_client: PyrogClient,
):
    for resource_id in mappings.keys():
        if mappings[resource_id]:
            continue
        mappings[resource_id] = pyrog_client.fetch_mapping(resource_id)


def batch(
    batch_instance: models.Batch,
    mappings: dict,
    topics_manager: TopicsManager,
    publisher: EventPublisher,
):
    for base_topic in ["batch", "extract", "transform", "load"]:
        topics_manager.create(f"{base_topic}.{batch_instance.id}")

    for resource_id in mappings.keys():
        publisher.publish(
            topic=f"batch.{batch_instance.id}",
            event=BatchEvent(batch_id=batch_instance.id, resource_id=resource_id),
        )


def abort(batch: models.Batch, topics_manager: TopicsManager) -> None:
    for base_topic in ["batch", "extract", "transform", "load"]:
        topics_manager.delete(f"{base_topic}.{batch.id}")

    batch.deleted_at = timezone.now()
    batch.save(update_fields=["deleted_at"])


def retry(batch: models.Batch) -> None:
    pass


def preview(
    resource_id: str, primary_key_values: Optional[list], pyrog_client: PyrogClient
) -> Tuple[List[Any], List[Any]]:
    resource_mapping = pyrog_client.fetch_mapping(resource_id)

    analyzer = Analyzer()
    analysis = analyzer.analyze(resource_mapping)

    db_connection = DBConnection(analysis.source_credentials)
    with db_connection.session_scope() as session:
        extractor = Extractor(session, db_connection.metadata)
        df = extractor.extract(analysis, primary_key_values)

        transformer = Transformer()

        documents = []
        errors = []

        for row in extractor.split_dataframe(df, analysis):
            # Encode and decode the row to mimic what happens when events are serialized
            # to pass through kafka
            row = json.JSONDecoder().decode(CustomJSONEncoder().encode(row))
            primary_key_value = row[analysis.primary_key_column.dataframe_column_name()][0]
            transformed_data = transformer.transform_data(row, analysis, primary_key_value)
            document = transformer.create_fhir_document(transformed_data, analysis, primary_key_value)
            documents.append(document)
            resource_type = document.get("resourceType")
            try:
                construct_fhir_element(resource_type, document)
            except ValidationError as e:
                errors.extend(
                    [
                        f"{err['msg'] or 'Validation error'}: "
                        f"{e.model.get_resource_type()}.{'.'.join([str(l) for l in err['loc']])}"
                        for err in e.errors()
                    ]
                )

    return documents, errors
