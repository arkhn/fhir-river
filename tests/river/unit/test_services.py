import uuid

import pytest

from river import models
from river.adapters.event_publisher import FakeEventPublisher
from river.adapters.mappings import FakeMappingsRepository
from river.adapters.topics import FakeTopics
from river.domain.events import BatchEvent
from river.services import abort, batch, preview

pytestmark = pytest.mark.django_db


def test_batch():
    topics = FakeTopics()
    publisher = FakeEventPublisher()
    resources = [str(uuid.uuid4()) for _ in range(5)]
    mappings_repo = FakeMappingsRepository(mappings={id: {} for id in resources})

    batch_instance = batch(resources, topics, publisher, mappings_repo)

    assert batch_instance is not None
    assert models.Batch.objects.filter(id=batch_instance.id).exists()
    assert topics._topics == {
        f"{base_topic}.{batch_instance.id}" for base_topic in ["batch", "extract", "transform", "load"]
    }
    assert publisher._events[f"batch.{batch_instance.id}"] == [
        BatchEvent(batch_id=batch_instance.id, resource_id=r) for r in resources
    ]
    assert mappings_repo._seen == resources


def test_abort(batch):
    topics = FakeTopics(
        topics=[f"{base_topic}.{batch.id}" for base_topic in ["batch", "extract", "transform", "load"]]
    )

    abort(batch, topics)

    assert batch.deleted_at is not None
    assert topics._topics == set()


@pytest.mark.skip(reason="feature not implemented yet")
def test_retry(batch):
    pass


def test_preview(users_to_patients_mapping):
    mappings_repo = FakeMappingsRepository(mappings={"foo": users_to_patients_mapping})

    documents, errors = preview("foo", None, mappings=mappings_repo)

    assert len(errors) == 0
