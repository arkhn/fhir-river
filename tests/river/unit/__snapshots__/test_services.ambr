# name: test_preview
  <class 'dict'> {
    'id': '21f20a6b-c3fa-5bd5-8fb2-4629315491c7',
    'identifier': <class 'list'> [
      <class 'dict'> {
        'system': 'http://terminology.arkhn.org/b8efd322-3e38-4072-9c68-e62e15d84d04',
        'value': '10006',
      },
    ],
    'meta': <class 'dict'> {
      'tag': <class 'list'> [
        <class 'dict'> {
          'code': 'cktlnp0bb00020mmz5fks20ij',
          'system': 'http://terminology.arkhn.org/CodeSystem/source',
        },
        <class 'dict'> {
          'code': 'cktlnp0f300300mmznmyqln70',
          'system': 'http://terminology.arkhn.org/CodeSystem/resource',
        },
      ],
    },
    'name': <class 'list'> [
      <class 'dict'> {
        'given': <class 'list'> [
          'Jean',
          'Georges',
        ],
      },
    ],
    'resourceType': 'Patient',
  }
---
